src/Microsoft.ML.Parquet/ParquetLoader.cs

```diff
@@ -390,11 +391,32 @@ public Cursor(ParquetLoader parent, Func<int, bool> predicate, IRandom rand)
                     Columns = _loader._columnsLoaded.Select(i => i.Name).ToArray()
                 };
 
-                int numBlocks = (int)Math.Ceiling(((decimal)parent.GetRowCount() / _readerOptions.Count));
-                int[] blockOrder = _rand == null ? Utils.GetIdentityPermutation(numBlocks) : Utils.GetRandomPermutation(rand, numBlocks);
-                _blockEnumerator = blockOrder.GetEnumerator();
+                // The number of blocks is calculated based on the specified rows in a block (defaults to 1M).
+                // Since we want to shuffle the blocks in addition to shuffling the rows in each block, checks
+                // are put in place to ensure we can produce a shuffle order for the blocks.
+                int numBlocks;
+                int[] blockOrder;
+                try
+                {
+                    numBlocks = checked((int)Math.Ceiling(((decimal)parent.GetRowCount() / _readerOptions.Count)));
```

Comment:
>decimal [](start = 60, length = 7)

why cast long to decimal? why not just stick to long?

Reply:
Because we want the `Math.Ceiling` to operate, whereas if we had this be a long it be the equiv to a floor -- that is, unless the number of rows happened to be exactly divisible by the count, then `numBlocks` would be 1 lower than it ought to be.

That said: I don't particularly like this cast to `decimal` myself and call to `Math` to solve what is a fairly straightforward problem. Elsewhere we solve the problem of `ceil((decimal)a / b)` by doing instead `(a + b -1) / b` on `int` or `long` types. That said that's not terribly obvious either, though at least it has the virtue of using familiar types. Perhaps an addition to `MathUtils`, since this sort of rounded-up-div problem comes up pretty often.

---
In reply to: [187491573](https://github.com/dotnet/machinelearning/pull/120#discussion_r187491573) [](ancestors = 187491573)

Reply:
Well, merely casing to long and doing a normal division won't save you. The issue isn't 2^^63 or any other large number. The issue is that divisions over integers are truncating divisions, which is the opposite of what we want sometimes like now when we're trying to figure out how many containers we'll need of capacity `c` if we have `r` items to store in them.

Let's consider, row count `long r=800` and `readerOptions.count`, let's call it `long c=100`. In this case, you can fit those into 8 blocks. `800/100 => 8`, `(800+100-1)/100=>8`, and `(long)Math.Ceiling(((decimal)800/100))=>8`.

Now consider if `r=801`. Suddenly we'll need `9` blocks. But `801/100 => 8`. Note however, `(801+101-1)/100=>9` and `(long)Math.Ceiling(((decimal)801/100))=>9`.

---
In reply to: [187506601](https://github.com/dotnet/machinelearning/pull/120#discussion_r187506601) [](ancestors = 187506601,187502257,187491573)

Reply:
Originally the decimal cast was because that's the only way to come out of the division with a decimal point instead of the closest lesser integer, but I've added a method in MathUtils to do the calculation with longs for clarity and to minimize casting instead.